cpp/src/io/orc/reader_impl.cu

```diff
@@ -633,6 +697,139 @@ void reader::impl::decode_stream_data(hostdevice_vector<gpu::ColumnDesc> &chunks
+  std::fill(num_child_rows.begin(), num_child_rows.end(), 0);
+  child_start_row.resize(number_of_child_chunks);
+  num_child_rows_per_stripe.resize(number_of_child_chunks);
+
+  int index = 0;  // number of child column processed
+
+  // For each parent column, update its child column meta for each stripe.
+  std::for_each(list_col.cbegin(), list_col.cend(), [&](auto p_col) {
+    const auto parent_col_idx = _col_meta.orc_col_map[level][p_col.id];
+    auto start_row            = 0;
+
+    for (size_t stripe_id = 0; stripe_id < number_of_stripes; stripe_id++) {
+      const auto child_rows = chunks[(stripe_id * num_parent_cols) + parent_col_idx].num_child_rows;
+
+      for (uint32_t id = 0; id < p_col.num_children; id++) {
+        const auto child_col_idx = index + id;
+
+        num_child_rows[child_col_idx] += child_rows;
+        num_child_rows_per_stripe[(stripe_id * num_child_cols) + child_col_idx] = child_rows;
+        // start row could be different for each column when there is nesting at each stripe level
+        child_start_row[(stripe_id * num_child_cols) + child_col_idx] =
+          (stripe_id == 0) ? 0 : start_row;
+      }
+      start_row += child_rows;
+    }
+    index += p_col.num_children;
+  });
+}
+
+std::unique_ptr<column> reader::impl::create_empty_column(int32_t orc_col_id,
+                                                          column_name_info &schema_info,
+                                                          rmm::cuda_stream_view stream)
+{
+  schema_info.name = _metadata->get_column_name(0, orc_col_id);
+  auto const type =
+    to_type_id(_metadata->get_schema(orc_col_id), _use_np_dtypes, _timestamp_type.id());
+  int32_t scale = 0;
+  std::vector<std::unique_ptr<column>> child_columns;
+  std::unique_ptr<column> out_col = nullptr;
+
+  switch (type) {
+    case type_id::LIST:
+      schema_info.children.emplace_back("offsets");
+      schema_info.children.emplace_back("");
+      out_col = make_lists_column(
+        0,
+        make_empty_column(data_type(type_id::INT32)),
+        create_empty_column(
+          _metadata->get_col_type(orc_col_id).subtypes[0], schema_info.children.back(), stream),
+        0,
+        rmm::device_buffer{0, stream},
+        stream);
+
+      break;
+
+    case type_id::STRUCT:
+      for (auto col : _metadata->get_col_type(orc_col_id).subtypes) {
+        schema_info.children.emplace_back("");
+        child_columns.push_back(create_empty_column(col, schema_info.children.back(), stream));
+      }
+      out_col =
+        make_structs_column(0, std::move(child_columns), 0, rmm::device_buffer{0, stream}, stream);
+      break;
+
+    case type_id::DECIMAL64:
+      scale = -static_cast<int32_t>(_metadata->get_types()[orc_col_id].scale.value_or(0));
+    default: out_col = make_empty_column(data_type(type, scale));
+  }
+
+  return out_col;
+}
+
+// Adds child column buffers to parent column
+column_buffer &&reader::impl::assemble_buffer(const int32_t orc_col_id,
```

Comment:
We are not creating new objects, it just updates children and for each parent buffer.

Reply:
but children are not references. We're moving objects into the `children` data member. This leaves "empty" `column_buffer` objects in the vector, AFAICT.

Reply:
I agree, but this vector is not used any more and I didn't want to create any more copies

Reply:
I think I get it. Then it would be good to pass `col_buffers` into `create_columns` as r-value ref, to declare that it's a sink and the object is not to be used after `create_columns`.